openshift/console
# OCPBUGS-16093,OCPBUGS-9409,OCPBUGS-16270: Creation of GH webhook and attaching it to repo while importing from git using PAC

**Fixes**: 
- https://issues.redhat.com/browse/OCPBUGS-16093
- https://issues.redhat.com/browse/OCPBUGS-9409
- https://issues.redhat.com/browse/OCPBUGS-16270

**Analysis / Root cause**: 
- [OCPBUGS-16093](https://issues.redhat.com/browse/OCPBUGS-16093): After the PAC Repository was created, we did not automate the step of adding the webhook to the remote Git repository. This was a manual step.
- [OCPBUGS-9409](https://issues.redhat.com/browse/OCPBUGS-9409): The local Bitbucket repositories were not reachable from the Import from Git form due to the CORS issue.
- [OCPBUGS-16270](https://issues.redhat.com/browse/OCPBUGS-16270): Info message below the "Git access token" Field for creating the Pipelines Repository under the Pipelines section in the Import from Git page is falling back to the default text instead of showing the curated ones of each Git provider.

**Solution Description**: 
- [OCPBUGS-16093](https://issues.redhat.com/browse/OCPBUGS-16093): Automated the process by using the Webhook creation APIs offered by GitHub, Bitbucket and Gitlab.
-  [OCPBUGS-9409](https://issues.redhat.com/browse/OCPBUGS-9409): Added a check that will ensure that any requests made to a local bitbucket repository will use the Proxy API that we have created in our backend. This will eliminate any CORS issues.
- [OCPBUGS-16270](https://issues.redhat.com/browse/OCPBUGS-16270): Updated the `gitProvider` values so that the info-text are visible in both the Git Import and PAC Repository Form.

**Screen shots / Gifs for design review**: 

1. [OCPBUGS-16093](https://issues.redhat.com/browse/OCPBUGS-16093):

https://github.com/openshift/console/assets/47265560/5dee32ac-169c-4829-856a-00a1992c33c8

**Complete Flow for triggering the PLRs upon Push & Pull Requests:**

https://github.com/openshift/console/assets/47265560/c54d33c5-92e1-4976-90ff-98fa7b72b63a

**Webhook Attachment Failure Toast:**
![image](https://github.com/openshift/console/assets/47265560/f35d2e79-82f5-4846-a8ad-9f0ff594493a)


2. [OCPBUGS-9409](https://issues.redhat.com/browse/OCPBUGS-9409): 

https://github.com/openshift/console/assets/47265560/c17c74fd-e76c-473b-949a-f34ca71ad7ff

3. [OCPBUGS-16270](https://issues.redhat.com/browse/OCPBUGS-16270):

https://github.com/openshift/console/assets/47265560/ff5884de-9047-4577-b9f9-54dce20912c8




**Unit test coverage report**: 
- Not changed.

**Test setup:**

1. [OCPBUGS-16093](https://issues.redhat.com/browse/OCPBUGS-16093):

- Go to the Import from Git Page:
- Clone the following repos based on the Git Provider you want to test:  
GH: https://github.com/Lucifergene/oc-pipe
GL: https://gitlab.com/avikkundu/oc-pipe
BB: https://bitbucket.org/avikkundu/oc-pipe
- Paste the cloned repo link in the Git URL field.
- Create the Secret Token following the instructions under the Secret Token Form field in the Pipelines Section of the form.
- Paste the Secret token and generate a WH secret.
- Click Create.
- If the token was created with all the required permissions, a webhook will be created in your GitHub repository. You can check it from this link: https://github.com/<username>/oc-pipe/settings/hooks
- If not, you will see the Webhook Creation Failed toast in the Topology Page giving the required info to create it manually.
- In the case of using the PAC repository form, on successful webhook creation, you will not see Step-4 in the Overview page, which normally shows the step of adding webhooks manually.

---

2. [OCPBUGS-9409](https://issues.redhat.com/browse/OCPBUGS-9409):

- Go to the Import from Git Page
- Paste a local BB Repository URL. ( e.g. https://open-bitbucket.nrao.edu/projects/PIPE/repos/pipeline/browse )
- Select BB as the Provider
- Check whether the URl is validated
- Create the Application

---

3. [OCPBUGS-16270](https://issues.redhat.com/browse/OCPBUGS-16270):

- Go to the Import from Git Page
- Paste the following links one by one in the Git URL field:
GH: https://github.com/Lucifergene/oc-pipe
GL: https://gitlab.com/avikkundu/oc-pipe
BB: https://bitbucket.org/avikkundu/oc-pipe
- For every URL, under the Git Secret Token Field under the Pipelines section, you will see specific info-texts for the selected Git Provider.



**Browser conformance**:
